### PR TITLE
Log Rotator

### PR DESCRIPTION
**Pull request by diptanu**
This is a library for rotating files written on a buffer. It will be used for log rotation of the stdout/stderr buffers by various drivers.

### CODE REVIEW

**Comment by pires** [comment]
All my Docker hosts have `logrotate` to fix what to me is a huge design flaw in Docker. So, this just makes a lot of sense. Thank you!

File: client/driver/logs.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"strings"
+)
+
+const (
+	bufSize = 32 * 1024 // Max number of bytes read from a buffer
+)
+
+// LogRotator rotates files for a buffer and retains only the last N rotated
+// files
+type LogRotator struct {
```
Comment (dadgar) [inline client/driver/logs.go]: I would update the comment to be something like:
```
// LogRotator ingests data and writes out to a rotated set of files
```

File: client/driver/logs.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+func NewLogRotator(path string, fileName string, maxFiles int, fileSize int64, logger *log.Logger) (*LogRotator, error) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	logFileIdx := 0
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), fileName) {
+			fileIdx := strings.TrimPrefix(f.Name(), fmt.Sprintf("%s.", fileName))
```
Comment (dadgar) [inline client/driver/logs.go]: Why not just check if it has this prefix: `fmt.Sprintf("%s.", fileName)`

File: client/driver/logs.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+// NewLogRotator configures and returns a new LogRotator
+func NewLogRotator(path string, fileName string, maxFiles int, fileSize int64, logger *log.Logger) (*LogRotator, error) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	logFileIdx := 0
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), fileName) {
```
Comment (dadgar) [inline client/driver/logs.go]: Why not just check if it has this prefix: `fmt.Sprintf("%s.", fileName)`

File: client/driver/logs.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+// LogRotator rotates files for a buffer and retains only the last N rotated
+// files
+type LogRotator struct {
+	maxFiles   int
+	fileSize   int64
+	path       string
+	fileName   string
+	logFileIdx int
```
Comment (dadgar) [inline client/driver/logs.go]: put comments on these fields

File: client/driver/logs.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+// NewLogRotator configures and returns a new LogRotator
+func NewLogRotator(path string, fileName string, maxFiles int, fileSize int64, logger *log.Logger) (*LogRotator, error) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	logFileIdx := 0
```
Comment (dadgar) [inline client/driver/logs.go]: Put a comment on what you are doing

File: client/driver/logs.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+				l.logFileIdx += 1
+				continue
+			}
+			remainingSize = l.fileSize - f.Size()
+		}
+		f, err := os.OpenFile(logFileName, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+		if err != nil {
+			return err
+		}
+		l.logger.Println("[INFO] logrotator: opened a new file: %s", logFileName)
```
Comment (dadgar) [inline client/driver/logs.go]: Not very useful message?

File: client/driver/logs.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		if remainingSize < 1 {
```
Comment (dadgar) [inline client/driver/logs.go]: `if remainingSize == 0 {`

More clear

File: client/driver/logs.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	}, nil
+}
+
+// Start reads from a Reader and writes them to files and rotates them when the
+// size of the file becomes equal to the max size configured
+func (l *LogRotator) Start(r io.Reader) error {
+	buf := make([]byte, bufSize)
+	for {
+		logFileName := filepath.Join(l.path, fmt.Sprintf("%s.%d", l.fileName, l.logFileIdx))
+		remainingSize := l.fileSize
```
Comment (dadgar) [inline client/driver/logs.go]: Put comments above each block of code describing what it does.

* Determine space left on this file
* Write logs to file until it is full

File: client/driver/logs.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+				return err
+			}
+			nw, err := f.Write(buf[:nr])
+			if err != nil {
+				f.Close()
+				return err
+			}
+			if nr != nw {
+				f.Close()
+				return fmt.Errorf("Failed to write data R: %d W: %d", nr, nw)
```
Comment (dadgar) [inline client/driver/logs.go]: Put a better error message

File: client/driver/logs.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+			}
+			remainingSize -= int64(nr)
+			if remainingSize < 1 {
```
Comment (dadgar) [inline client/driver/logs.go]: Same here

File: client/driver/logs.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	count := 0
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), l.fileName) {
+			fileIdx := strings.TrimPrefix(f.Name(), fmt.Sprintf("%s.", l.fileName))
+			n, err := strconv.Atoi(fileIdx)
+			if err != nil {
+				continue
+			}
+			if count == l.maxFiles {
```
Comment (dadgar) [inline client/driver/logs.go]: Why not move this outside the loop and then you can iterator over a sub-slice calling delete. So you only sort once

File: client/driver/logs.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+				sort.Sort(sort.Reverse(sort.IntSlice(fIndexes)))
+				fname := filepath.Join(l.path, fmt.Sprintf("%s.%d", l.fileName, fIndexes[count-1]))
+				l.logger.Printf("[INFO] removing file: %v", fname)
```
Comment (dadgar) [inline client/driver/logs.go]: Probably un-needed too.

File: client/driver/logs_test.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+	"testing"
+)
+
+var (
+	logger = log.New(os.Stdout, "", log.LstdFlags)
+	path   = "/tmp/logrotator"
+)
+
+func TestLogRotator_IncorrectPath(t *testing.T) {
+	incorrectPath := "/foo"
```
Comment (dadgar) [inline client/driver/logs_test.go]: incorrect makes it seem like you are expecting it to be something. I would name it emptyPath or invalidPath

File: client/driver/logs_test.go [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+	if _, err := NewLogRotator(incorrectPath, "redis.stdout", 10, 10, logger); err == nil {
+		t.Fatal("expected err")
+	}
+}
+
+func TestLogRotator_FindCorrectIndex(t *testing.T) {
+	if err := os.Mkdir(path, os.ModeDir|os.ModePerm); err != nil {
```
Comment (dadgar) [inline client/driver/logs_test.go]: https://golang.org/pkg/io/ioutil/#TempDir

File: client/driver/logs_test.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+		t.Fatalf("test setup err: %v", err)
+	}
+	fname := filepath.Join(path, "redis.stdout.0")
+	if f, err := os.Create(fname); err == nil {
+		f.WriteString("abcde")
+		f.Close()
+	}
+
+	l, err := NewLogRotator(path, "redis.stdout", 10, 6, logger)
+	if err != nil && err != io.EOF {
```
Comment (dadgar) [inline client/driver/logs_test.go]: is io.EOF expected here?

File: client/driver/logs_test.go [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+		t.Fatalf("test setup err: %v", err)
+	}
+
+	r, w := io.Pipe()
+	go func() {
+		w.Write([]byte("fg"))
```
Comment (dadgar) [inline client/driver/logs_test.go]: Put a comment that this should roll over

File: client/driver/logs_test.go [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+		t.Fatalf("Failure in logrotator start %v", err)
+	}
+
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	numFiles := 0
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), "redis.stdout.") {
```
Comment (dadgar) [inline client/driver/logs_test.go]: I would just read the other file and check it has the byte

File: client/driver/logs_test.go (hunk 15 above)
Comment (diptanu) [inline client/driver/logs_test.go]: @dadgar Yes, when we close the pipe.

**Pull request merged**